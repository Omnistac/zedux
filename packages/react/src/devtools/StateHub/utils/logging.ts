import { AnyAtomInstance } from '@zedux/react'

export const logAtomInstance = (instance: AnyAtomInstance, state: any) =>
  logGroup(
    `Atom Instance ${instance.keyHash}`,
    'Instance:',
    instance,
    'State:',
    state
  )

export const logGroup = (title: string, ...args: any[]) => {
  console.group(title)
  args.forEach(arg => {
    console.log(arg)
  })
  console.groupEnd()
}
