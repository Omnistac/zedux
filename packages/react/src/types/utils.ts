import { AtomTemplateBase } from '../classes/templates/AtomTemplateBase'
import { AtomInstance } from '../classes/instances/AtomInstance'
import { AtomInstanceBase } from '../classes/instances/AtomInstanceBase'

export type AnyAtomInstance = AtomInstance<any>
export type AnyAtomInstanceBase = AtomInstanceBase<
  any,
  AtomTemplateBase<any, AtomInstance<any>>
>
export type AnyAtomTemplate = AtomTemplateBase<any, any>
