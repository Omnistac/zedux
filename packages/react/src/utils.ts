import { AnyAtomTemplate, Ecosystem } from '@zedux/atoms'
import React, { createContext } from 'react'

export const ecosystemContext = createContext('@@global')

/**
 * These are copied from packages/atoms/src/utils/general.ts
 *
 * IMPORTANT: keep these in-sync with the ones in the atoms package
 */
export const Explicit = 1
export const External = 2
export const Static = 4
// export const Deferred = 8

export const destroyed = Symbol.for(`@@zedux/destroyed`)

export const getReactContext = (
  ecosystem: Ecosystem,
  atom: AnyAtomTemplate
) => {
  const reactStorage: {
    contexts: WeakMap<typeof createContext, Record<string, React.Context<any>>>
  } = (ecosystem._storage.react ||= {})

  const contexts = (reactStorage.contexts ||= new WeakMap())
  const windowContexts =
    contexts.get(createContext) ||
    (contexts.set(createContext, {}).get(createContext) as Record<
      string,
      React.Context<any>
    >)

  return (windowContexts[atom.key] ||= createContext(
    undefined
  ) as React.Context<any>)
}
