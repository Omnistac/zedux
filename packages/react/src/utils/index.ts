export * from './general'
export * from './types'
