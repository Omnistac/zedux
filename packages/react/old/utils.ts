import { Atom, Scope } from './types'

let idCounter = 0

const generateId = () => {
  idCounter += 1
  return idCounter
}

const stringifyParam = (param: any): string => {
  if (Array.isArray(param)) {
    return `[${param.map(child => stringifyParam(child)).join(',')}]`
  }

  if (param && typeof param === 'object') {
    return `{${Object.entries(param)
      .map(([key, val]) => `${key}: ${stringifyParam(val)}`)
      .sort((a, b) => a.localeCompare(b))
      .join(',')}}`
  }

  return JSON.stringify(param)
}

export const generateAppId = () => `app-${generateId()}`
export const generateImplementationId = () => `im-${generateId()}`
export const generateInstanceId = () => `in-${generateId()}`
export const generateLocalId = () => `lo-${generateId()}`

export const getFullKey = (atom: Atom, params?: any[]) => {
  if (atom.scope === Scope.local) return generateLocalId()

  if (!params?.length) return atom.key

  return `${atom.key}---${stringifyParam(params)}`
}
