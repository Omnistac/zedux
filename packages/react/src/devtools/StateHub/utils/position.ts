import { GridNum, GridProps, Size } from '../types'
import { css } from '../simple-styled-components'

const sizeMap: Record<Size, number> = {
  0: 0.8,
  1: 1.3,
  2: 2.5,
  3: 7,
  4: 16,
  5: 1, // 5 is full screen anyway
}

export const getGridColumn = (props: GridProps) => {
  const isWide = getIsWide(props)

  return isWide ? '1 / span 3' : `${props.col + 1} / span 1`
}

export const getGridRow = (props: GridProps) => {
  const isTall = getIsTall(props)

  return isTall ? '1 / span 3' : `${props.row + 1} / span 1`
}

export const getIsWide = ({ col, row }: GridProps) =>
  (row === 0 && col === 1) || (row === 2 && col === 1)

export const getIsTall = ({ col, row }: GridProps) =>
  row === 1 && (col === 0 || col === 2)

export const getTemplate = (row: GridNum, col: GridNum, size: Size) => {
  const fillStr = size === 5 ? '0' : '1fr'
  const rows = Array(3).fill(fillStr)
  const cols = Array(3).fill(fillStr)
  const isTall = size === 5 ? false : getIsTall({ col, row })
  const isWide = size === 5 ? false : getIsWide({ col, row })
  const mappedSize = sizeMap[size]

  cols[col] = isTall ? `${mappedSize - 0.4}fr` : `${mappedSize}fr`
  rows[row] = isWide ? `${mappedSize - 0.4}fr` : `${mappedSize}fr`

  return css`
    grid-template-columns: ${cols.join(' ')};
    grid-template-rows: ${rows.join(' ')};
  `
}
