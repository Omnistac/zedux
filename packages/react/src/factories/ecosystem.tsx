import { Ecosystem } from '../classes/Ecosystem'
import { EcosystemConfig } from '../types'

export const ecosystem = (config: EcosystemConfig) => new Ecosystem(config)
