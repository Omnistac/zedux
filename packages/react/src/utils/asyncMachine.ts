import {
  createActorFactory,
  createMachine,
  createStore,
  states,
} from '@zedux/core'
import { AsyncStatus } from '../types'

const createActor = createActorFactory('@zedux/react/asyncMachine')
export const cancel = createActor('cancel')
export const load = createActor('load')
export const loadSuccess = createActor<any>('loadSuccess')
export const loadError = createActor<Error>('loadError')
export const reset = createActor('reset')

export const [errorState, idleState, loadingState, successState] = states(
  AsyncStatus.Error,
  AsyncStatus.Idle,
  AsyncStatus.Loading,
  AsyncStatus.Success
)

export const asyncMachine = createMachine<AsyncStatus>(
  idleState.on(load, loadingState),
  loadingState
    .on(loadSuccess, successState)
    .on(loadError, errorState)
    .on(cancel, idleState),
  successState.on(reset, idleState),
  errorState.on(reset, idleState)
)

export const createAsyncMachineStore = () => createStore(asyncMachine)
