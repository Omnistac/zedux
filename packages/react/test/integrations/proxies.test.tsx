import { ecosystem } from '../utils/ecosystem'

describe('proxies', () => {
  test('set operations', () => {
    const calls: any[] = []
    const initialObj = { a: [new Set('b')], id: 1 }
    const signal = ecosystem.signal(new Set([initialObj]))

    signal.on('mutate', (transactions, { change }) => {
      calls.push([transactions, change?.newState])
    })

    signal.mutate(state => {
      state.add({ a: [], id: 2 })
    })

    const expectedState1 = new Set([initialObj, { a: [], id: 2 }])

    expect(signal.get()).toEqual(expectedState1)

    expect(calls).toEqual([
      [[{ k: { a: [], id: 2 }, v: undefined }], expectedState1],
    ])
    calls.splice(0, calls.length)

    signal.mutate(state => {
      state.delete(initialObj)
    })

    const expectedState2 = new Set([{ a: [], id: 2 }])

    expect(signal.get()).toEqual(expectedState2)

    expect(calls).toEqual([[[{ k: initialObj, t: 'd' }], expectedState2]])
    calls.splice(0, calls.length)

    signal.mutate(state => state.clear())

    expect(signal.get()).toEqual(new Set())
    expect(calls).toEqual([[[{ k: { a: [], id: 2 }, t: 'd' }], new Set()]])
    calls.splice(0, calls.length)

    signal.mutate(state => {
      if (state.has(initialObj) || state.size === 0) state.delete(initialObj)
    })

    expect(signal.get()).toEqual(new Set())
    expect(calls).toEqual([])
  })

  test('array operations', () => {
    const calls: any[] = []
    const signal = ecosystem.signal([{ a: [1, 3] }])

    signal.on('mutate', (transactions, { change }) => {
      calls.push([transactions, change?.newState])
    })

    signal.mutate([{ a: [11, 33] }])

    expect(signal.get()).toEqual([{ a: [11, 33] }])

    signal.mutate(state => {
      state.push({ a: [2] }, { a: [200] })
    })

    expect(signal.get()).toEqual([{ a: [11, 33] }, { a: [2] }, { a: [200] }])

    signal.mutate(state => {
      state.splice(1, 1)
    })

    expect(signal.get()).toEqual([{ a: [11, 33] }, { a: [200] }])

    signal.mutate(state => {
      state.unshift({ a: [3] })
      state.splice(1, 1, { a: [300] })
    })

    expect(signal.get()).toEqual([{ a: [3] }, { a: [300] }, { a: [200] }])

    signal.mutate(state => {
      if (state.length > 2) {
        state.shift()
        state[1].a.push(201)
        state[0].a = []
      }
    })

    expect(signal.get()).toEqual([{ a: [] }, { a: [200, 201] }])

    signal.mutate(state => {
      const val = state.pop()
      state.splice(0, 0, val!)

      // if it's valid JS, it's a valid mutation/transaction
      state.splice(10, 10, { a: [400] })
    })

    expect(signal.get()).toEqual([{ a: [200, 201] }, { a: [] }, { a: [400] }])

    signal.mutate(state => {
      const symbol = state[Symbol.iterator]
      expect(symbol).toBe(Array.prototype[Symbol.iterator])
    })

    expect(calls).toMatchSnapshot()
  })

  test('object operations', () => {
    const calls: any[] = []
    const signal = ecosystem.signal({ a: { b: 'b' }, c: null as any })

    signal.on('mutate', (transactions, { change }) => {
      calls.push([transactions, change?.newState])
    })

    signal.mutate(state => ({ a: { b: state.a.b + 'b' } }))

    expect(signal.get()).toEqual({ a: { b: 'bb' }, c: null })

    signal.mutate({ c: { d: 'd' } })

    expect(signal.get()).toEqual({ a: { b: 'bb' }, c: { d: 'd' } })

    signal.mutate(state => {
      delete state.c
    })

    expect(signal.get()).toEqual({ a: { b: 'bb' } })

    signal.mutate(state => {
      state.c = undefined
      state.a.b += 'b'
    })

    expect(signal.get()).toEqual({ a: { b: 'bbb' }, c: undefined })

    expect(calls).toMatchSnapshot()
  })

  test('unsupported operations throw', () => {
    const signal = ecosystem.signal(['a', 'z', 'b', 'y'])
    const pattern = /This operation is not supported/

    expect(() => signal.mutate(state => state.reverse())).toThrow(pattern)

    expect(() => signal.mutate(state => state.sort())).toThrow(pattern)

    expect(() =>
      signal.mutate(state => {
        delete state[0]
      })
    ).toThrow(pattern)

    const setSignal = ecosystem.signal(new Set())

    expect(() =>
      setSignal.mutate(state => {
        // @ts-expect-error can't assign to a set
        state.nothing = 'nothing'
      })
    ).toThrow(pattern)
  })

  test('Zedux recursively proxies a returned object only if no transactions were added', () => {
    const calls: any[] = []
    const signal = ecosystem.signal({ a: { b: 'b' } })

    signal.on('mutate', (transactions, { change }) => {
      calls.push([transactions, change?.newState])
    })

    signal.mutate(state => {
      state.a.b = 'bb'

      return { a: { b: 'ignored' } }
    })

    expect(signal.get()).toEqual({ a: { b: 'bb' } })
    expect(calls).toEqual([[[{ k: ['a', 'b'], v: 'bb' }], { a: { b: 'bb' } }]])
    calls.splice(0, calls.length)

    signal.mutate(state => {
      state.a.b

      return { a: { b: 'not ignored' } }
    })

    expect(signal.get()).toEqual({ a: { b: 'not ignored' } })
    expect(calls).toEqual([
      [[{ k: ['a', 'b'], v: 'not ignored' }], { a: { b: 'not ignored' } }],
    ])
  })
})
