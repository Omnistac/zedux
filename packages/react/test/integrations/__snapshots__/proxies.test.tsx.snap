// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`proxies array operations 1`] = `
[
  [
    [
      {
        "k": [
          "0",
          "a",
          "0",
        ],
        "v": 11,
      },
      {
        "k": [
          "0",
          "a",
          "1",
        ],
        "v": 33,
      },
    ],
    [
      {
        "a": [
          11,
          33,
        ],
      },
    ],
  ],
  [
    [
      {
        "k": 1,
        "v": {
          "a": [
            2,
          ],
        },
      },
      {
        "k": 2,
        "v": {
          "a": [
            200,
          ],
        },
      },
    ],
    [
      {
        "a": [
          11,
          33,
        ],
      },
      {
        "a": [
          2,
        ],
      },
      {
        "a": [
          200,
        ],
      },
    ],
  ],
  [
    [
      {
        "k": 1,
        "t": "d",
      },
    ],
    [
      {
        "a": [
          11,
          33,
        ],
      },
      {
        "a": [
          200,
        ],
      },
    ],
  ],
  [
    [
      {
        "k": 0,
        "t": "i",
        "v": {
          "a": [
            3,
          ],
        },
      },
      {
        "k": 1,
        "t": "d",
      },
      {
        "k": 1,
        "t": "i",
        "v": {
          "a": [
            300,
          ],
        },
      },
    ],
    [
      {
        "a": [
          3,
        ],
      },
      {
        "a": [
          300,
        ],
      },
      {
        "a": [
          200,
        ],
      },
    ],
  ],
  [
    [
      {
        "k": 0,
        "t": "d",
      },
      {
        "k": [
          "1",
          "a",
          1,
        ],
        "v": 201,
      },
      {
        "k": [
          "0",
          "a",
        ],
        "v": [],
      },
    ],
    [
      {
        "a": [],
      },
      {
        "a": [
          200,
          201,
        ],
      },
    ],
  ],
  [
    [
      {
        "k": 1,
        "t": "d",
      },
      {
        "k": 0,
        "t": "i",
        "v": {
          "a": [
            200,
            201,
          ],
        },
      },
      {
        "k": 10,
        "t": "i",
        "v": {
          "a": [
            400,
          ],
        },
      },
    ],
    [
      {
        "a": [
          200,
          201,
        ],
      },
      {
        "a": [],
      },
      {
        "a": [
          400,
        ],
      },
    ],
  ],
]
`;

exports[`proxies object operations 1`] = `
[
  [
    [
      {
        "k": [
          "a",
          "b",
        ],
        "v": "bb",
      },
    ],
    {
      "a": {
        "b": "bb",
      },
      "c": null,
    },
  ],
  [
    [
      {
        "k": "c",
        "v": {
          "d": "d",
        },
      },
    ],
    {
      "a": {
        "b": "bb",
      },
      "c": {
        "d": "d",
      },
    },
  ],
  [
    [
      {
        "k": "c",
        "t": "d",
      },
    ],
    {
      "a": {
        "b": "bb",
      },
    },
  ],
  [
    [
      {
        "k": "c",
        "v": undefined,
      },
      {
        "k": [
          "a",
          "b",
        ],
        "v": "bbb",
      },
    ],
    {
      "a": {
        "b": "bbb",
      },
      "c": undefined,
    },
  ],
]
`;
